src/object_gui.cpp

```diff
@@ -72,42 +87,127 @@ class BuildObjectWindow : public Window {
 	}
 
 public:
-	BuildObjectWindow(WindowDesc *desc, WindowNumber number) : Window(desc), info_height(1)
+	BuildObjectWindow(WindowDesc *desc, WindowNumber number) : Window(desc), info_height(1), filter_editbox(EDITBOX_MAX_SIZE)
 	{
 		this->CreateNestedTree();
+
 		this->vscroll = this->GetScrollbar(WID_BO_SCROLLBAR);
-		this->FinishInitNested(number);
 
-		ResetObjectToPlace();
+		this->querystrings[WID_BO_FILTER] = &this->filter_editbox;
 
-		this->vscroll->SetPosition(0);
-		this->vscroll->SetCount(ObjectClass::GetUIClassCount());
+		this->object_classes.SetListing(this->last_sorting);
+		this->object_classes.SetFiltering(this->last_filtering);
+		this->object_classes.SetSortFuncs(this->sorter_funcs);
+		this->object_classes.SetFilterFuncs(this->filter_funcs);
+		this->object_classes.ForceRebuild();
+
+		BuildObjectClassesAvailable();
+		SelectClassAndObject();
+
+		this->FinishInitNested(number);
 
 		NWidgetMatrix *matrix = this->GetWidget<NWidgetMatrix>(WID_BO_SELECT_MATRIX);
 		matrix->SetScrollbar(this->GetScrollbar(WID_BO_SELECT_SCROLL));
 
-		this->SelectOtherClass(_selected_object_class);
+		this->SetFocusedWidget(WID_BO_FILTER);
+		this->GetWidget<NWidgetMatrix>(WID_BO_OBJECT_MATRIX)->SetCount(4);
+
+		ResetObjectToPlace();
+
+		this->vscroll->SetCount((int) this->object_classes.size());
+
+		EnsureSelectedObjectClassIsVisible();
+
+		this->InvalidateData();
+	}
+
+	/** Sort object classes by class name. */
+	static bool NameSorter(const ObjectClass *const& a, const ObjectClass *const& b)
+	{
+		char buffer_a[DRAW_STRING_BUFFER];
+		GetString(buffer_a, a->name, lastof(buffer_a));
+
+		char buffer_b[DRAW_STRING_BUFFER];
+		GetString(buffer_b, b->name, lastof(buffer_b));
+
+		return strnatcmp(buffer_a, buffer_b, true) < 0; // Sort by name (natural sorting).
+	}
+
+	/** Filter object classes by class name. */
+	static bool CDECL TagNameFilter(const ObjectClass *const* oc, StringFilter& filter)
+	{
+		char buffer[DRAW_STRING_BUFFER];
+		GetString(buffer, (*oc)->name, lastof(buffer));
+
+		filter.ResetState();
+		filter.AddLine(buffer);
+		return filter.GetState();;
+	}
+
+	void BuildObjectClassesAvailable()
+	{
+		if (!this->object_classes.NeedRebuild()) return;
+
+		this->object_classes.clear();
+
+		for (uint i = 0; i < ObjectClass::GetClassCount(); i++) {
+			ObjectClass *objclass = ObjectClass::Get((ObjectClassID)i);
+			if (objclass->GetUISpecCount() == 0) continue; // Is this needed here?
+			this->object_classes.push_back(objclass);
+		}
+
+		this->object_classes.Filter(this->string_filter);
+		this->object_classes.shrink_to_fit();
+		this->object_classes.RebuildDone();
+		this->object_classes.Sort();
+
+		this->vscroll->SetCount((uint)this->object_classes.size());
+	}
+
+	void SelectClassAndObject()
+	{
+		assert(!this->object_classes.empty()); // object GUI should be disabled elsewise
+		if (_selected_object_class == nullptr) {
+			// This happens during the first time the window is open during the app life cycle.
```

Comment:
According to [coding style](https://wiki.openttd.org/en/Development/Coding%20style#other-important-rules):
```suggestion
			/* This happens during the first time the window is open during the app life cycle. */
```